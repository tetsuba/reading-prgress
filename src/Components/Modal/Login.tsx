import { useMutation } from 'react-query'
import { useNavigate } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { formDataToObject, getErrorMessage } from '../../lib/utils'
import { updateUser } from '../../store/user/userSlice'
import { addBooks } from '../../store/books/booksSlice'

// COMPONENTS
import H3 from '../H3/H3'
import Label from '../Form/Label'
import Input from '../Form/Input'
import ErrorMessage from '../Form/ErrorMessage'
import Button from '../Button/Button'
import { loginUser } from '../../api/user'
import { addStudents } from '../../store/students/studentsSlice'
import { useEffect, useRef } from 'react'

type PropTypes = {
    setShowLogin: (p: boolean) => void
}
export default function Login(props: PropTypes) {
    const eleRefButton = useRef<HTMLButtonElement | null>(null)
    const dispatch = useDispatch()
    const navigate = useNavigate()
    const mutation = useMutation(loginUser, {
        onSuccess: (data) => {
            const {
                data: { user, token, books, students }
            } = data
            dispatch(updateUser({ user, token }))
            dispatch(addBooks(books))
            dispatch(addStudents(students))
            props.setShowLogin(false)
            navigate('/dashboard')
        }
    })

    useEffect(() => {
        // TODO: Fix this hacky way of focusing on the first input element after the modal is rendered
        setTimeout(() => {
            if (eleRefButton.current !== null) {
                eleRefButton.current.focus()
            }
        }, 0)
    }, [])

    return (
        <div
            role="dialog"
            aria-labelledby="Login"
            aria-describedby="Please enter your username and password"
            data-testid="login-view"
        >
            <Button
                tabIndex={0}
                ref={eleRefButton}
                className="absolute right-2.5 top-3"
                data-testid="modal-close"
                icon="close"
                onClick={() => props.setShowLogin(false)}
                template="icon-close"
            />
            <div className="px-6 py-6 lg:px-8">
                <H3 className="mb-3">Sign in to our platform</H3>
                <form
                    data-testid="login-form"
                    className="space-y-6"
                    onSubmit={(event) => {
                        event.preventDefault()
                        const obj = formDataToObject(event.target)
                        mutation.mutate(obj)
                    }}
                >
                    <div>
                        <Label htmlFor="email">Your email</Label>
                        <Input
                            template="text"
                            type="email"
                            name="email"
                            placeholder="name@company.com"
                        />
                    </div>
                    <div>
                        <Label htmlFor="password">Your password</Label>
                        <Input
                            template="text"
                            type="password"
                            name="password"
                            placeholder="••••••••"
                        />
                    </div>
                    <div className="flex justify-between">
                        <div className="flex items-start">
                            <div className="flex h-5 items-center">
                                <Input
                                    template="checkbox"
                                    type="checkbox"
                                    id="remember-me"
                                    value=""
                                />
                            </div>
                            <Label htmlFor="remember" className="ml-1">
                                Remember me
                            </Label>
                        </div>
                        <Button
                            template="text-link"
                            onClick={() => {
                                navigate('/lostPassword')
                                props.setShowLogin(false)
                            }}
                        >
                            Lost Password?
                        </Button>
                    </div>
                    <ErrorMessage
                        show={mutation.isError}
                        className="mt-1 block"
                    >
                        {getErrorMessage(mutation.error as Error)}
                    </ErrorMessage>
                    <Button
                        template="primary"
                        className="w-full justify-center"
                        type="submit"
                    >
                        Login to your account
                    </Button>
                    <div className="text-sm font-medium text-gray-500 dark:text-gray-300">
                        Not registered?
                        <Button
                            onBlur={() => eleRefButton.current?.focus()}
                            className="ml-1"
                            template="text-link"
                            onClick={() => {
                                navigate('/register')
                                props.setShowLogin(false)
                            }}
                        >
                            Create account
                        </Button>
                    </div>
                </form>
            </div>
        </div>
    )
}
