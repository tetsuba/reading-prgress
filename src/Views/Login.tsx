import {useMutation} from "react-query"
import {formDataToObject, getErrorMessage} from "../lib/utils"
import {loginUser} from "../lib/service"

// COMPONENTS
import Button from "../Components/Button/Button"
import H3 from "../Components/H3/H3"
import Label from "../Components/Form/Label"
import Input from "../Components/Form/Input"
import ErrorMessage from "../Components/Form/ErrorMessage";

type PropTypes = {
    setShowLogin: (p: boolean) => void
}
export default function Login (props: PropTypes) {
    const mutation = useMutation(loginUser)

    if (mutation.isSuccess) {
        props.setShowLogin(false)
    }

    return (
        <div data-testid="login-view">
            <Button
                dataTestid="modal-close"
                type="button"
                svg="close"
                clickHandler={() => props.setShowLogin(false)}
            />
            <div className="px-6 py-6 lg:px-8">
                <H3 className="mb-3">Sign in to our platform</H3>
                <form
                    data-testid="login-form"
                    className="space-y-6"
                    onSubmit={(event) => {
                        event.preventDefault()
                        const obj = formDataToObject(event.target)
                        mutation.mutate(obj)
                    }}
                >
                    <div>
                        <Label htmlFor="email">Your email</Label>
                        <Input type="email" name="email" placeholder="name@company.com" />
                    </div>
                    <div>
                        <Label htmlFor="password">Your password</Label>
                        <Input type="password" name="password" placeholder="••••••••" />
                    </div>
                    <div className="flex justify-between">
                        <div className="flex items-start">
                            <div className="flex items-center h-5">
                                <Input type="checkbox" id="remember-me" value="" />
                            </div>
                            <Label htmlFor="remember" className="ml-1">Remember me</Label>
                        </div>
                        <a href="src/Components/Modal/View#" className="text-sm text-blue-700 hover:underline dark:text-blue-500">Lost Password?</a>
                    </div>
                    <ErrorMessage show={mutation.isError} className="mt-1 block">
                        {getErrorMessage(mutation.error)}
                    </ErrorMessage>
                    <Button type="submit">Login to your account</Button>
                    <div className="text-sm font-medium text-gray-500 dark:text-gray-300">
                        Not registered?
                        <a href="src/Components/Modal/View#" className="text-blue-700 hover:underline dark:text-blue-500">Create account</a>
                    </div>
                </form>
            </div>
        </div>
    )
}

