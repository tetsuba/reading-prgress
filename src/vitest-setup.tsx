import { QueryClient, QueryClientProvider } from 'react-query'
import { MemoryRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import store from './store/store'
import { updateUser } from './store/user/userSlice'
import { addBooks } from './store/books/booksSlice'
import { addStudents } from './store/students/studentsSlice'
import { studentsMockData } from '../tests/mockData/students'
import booksMockData from '../tests/mockData/books'
import { userMockData } from '../tests/mockData/user'

const portal = document.createElement('div')
portal.setAttribute('id', 'modal')
document.body.appendChild(portal)

const mockResults = {
    results: [[{ transcript: 'This is a story' }]]
}

class MockSpeech {
    start() {
        // @ts-ignore
        this.onresult(mockResults)
    }
    stop = vi.fn()
    onresult = vi.fn()
    onend = vi.fn()
}

Object.defineProperty(window, 'webkitSpeechRecognition', {
    writable: true,
    value: vi.fn().mockImplementation(() => new MockSpeech())
})

const queryClient = new QueryClient({
    defaultOptions: {
        queries: {
            staleTime: 0,
            cacheTime: 0,
            retryDelay: 1,
            retry: 0
        }
    }
})

type PropTypes = {
    pathname: string
    children: JSX.Element
    notAuthenticated?: boolean
}

export function updateStoreWithMockData() {
    store.dispatch(updateUser(userMockData))
    store.dispatch(addBooks(booksMockData))
    store.dispatch(addStudents(studentsMockData))
}

export function WrapperWith_Store_Query_Router(props: PropTypes) {
    if (!props.notAuthenticated) updateStoreWithMockData()
    return (
        <Provider store={store}>
            <QueryClientProvider client={queryClient}>
                <MemoryRouter initialEntries={[props.pathname]}>
                    {props.children}
                </MemoryRouter>
            </QueryClientProvider>
        </Provider>
    )
}

export function WrapperWith_Store_Router(props: PropTypes) {
    return (
        <Provider store={store}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </Provider>
    )
}

export function WrapperWithQueryAndRouter(props: PropTypes) {
    return (
        <QueryClientProvider client={queryClient}>
            <MemoryRouter initialEntries={[props.pathname]}>
                {props.children}
            </MemoryRouter>
        </QueryClientProvider>
    )
}

export function WrapperWithQuery(props: { children: JSX.Element }) {
    return (
        <QueryClientProvider client={queryClient}>
            {props.children}
        </QueryClientProvider>
    )
}

export function WrapperWithRouter(props: PropTypes) {
    return (
        <MemoryRouter initialEntries={[props.pathname]}>
            {props.children}
        </MemoryRouter>
    )
}
