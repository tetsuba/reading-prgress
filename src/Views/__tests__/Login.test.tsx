import { fireEvent, render, screen, waitFor, act } from '@testing-library/react'
import { Mocked } from 'vitest'
import { WrapperWith_Store_Query_Router } from '../../vitest-setup'
import Login from '../../Components/Modal/Login'

import axios from 'axios'
vi.mock('axios')
const mockedAxios = axios as Mocked<typeof axios>

/* Note:
 * This was mocked to stop a warning message from being displayed
 * when running the test.
 * */
const mockNavigate = vi.fn()
vi.mock('react-router-dom', () => ({
    ...(vi.importActual('react-router-dom') as object),
    useNavigate: () => mockNavigate
}))

const setShowLoginMock = vi.fn()

describe('Login', () => {
    test('should render', () => {
        const { asFragment } = render(
            <WrapperWith_Store_Query_Router pathname="/">
                <Login setShowLogin={setShowLoginMock} />
            </WrapperWith_Store_Query_Router>
        )
        expect(asFragment()).toMatchSnapshot()
    })
    test('should return an error if email or password is incorrect', async () => {
        const AXIOS_ERROR: any = {
            code: 'ERR_BAD_RESPONSE',
            message: 'Request failed with status code 500',
            name: 'AxiosError',
            status: 500,
            statusText: 'Internal Server Error',
            response: {
                data: { error: 'Incorrect username or password' },
                status: 500
            }
        }

        mockedAxios.post.mockRejectedValueOnce(AXIOS_ERROR)

        const eventTarget = {
            target: {
                password: {
                    value: '1234'
                },
                email: {
                    value: 'dsjfkldsf@fjkdlsfj.com'
                }
            }
        }

        render(
            <WrapperWith_Store_Query_Router pathname="/">
                <Login setShowLogin={setShowLoginMock} />
            </WrapperWith_Store_Query_Router>
        )

        act(() => {
            fireEvent.submit(screen.getByTestId('login-form'), eventTarget)
        })

        await waitFor(() => expect(screen.getByTestId('error-message')))
    })
    test('should return success with correct login details', async () => {
        mockedAxios.post.mockResolvedValue({
            data: {
                user: {
                    firstName: 'firstName',
                    lastName: 'lastName',
                    email: 'email',
                    id: 'id'
                }
            }
        })

        const eventTarget = {
            target: {
                password: {
                    value: '123456'
                },
                email: {
                    value: 'dsjfkldsf@fjkdlsfj.com'
                }
            }
        }

        render(
            <WrapperWith_Store_Query_Router pathname="/">
                <Login setShowLogin={setShowLoginMock} />
            </WrapperWith_Store_Query_Router>
        )

        act(() => {
            fireEvent.submit(screen.getByTestId('login-form'), eventTarget)
        })

        await waitFor(() => expect(setShowLoginMock).toHaveBeenCalled())
    })
    test('should click on the lost password text link', () => {
        render(
            <WrapperWith_Store_Query_Router pathname="/">
                <Login setShowLogin={setShowLoginMock} />
            </WrapperWith_Store_Query_Router>
        )

        fireEvent.click(screen.getByText('Lost Password?'))
        expect(mockNavigate).toHaveBeenCalled()
    })
    test('should click on the create and account text link', () => {
        render(
            <WrapperWith_Store_Query_Router pathname="/">
                <Login setShowLogin={setShowLoginMock} />
            </WrapperWith_Store_Query_Router>
        )

        fireEvent.click(screen.getByText('Create account'))
        expect(mockNavigate).toHaveBeenCalled()
    })
})
