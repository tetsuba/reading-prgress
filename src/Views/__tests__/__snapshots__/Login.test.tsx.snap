// Vitest Snapshot v1

exports[`Login > should render 1`] = `
<DocumentFragment>
  <div
    data-testid="login-view"
  >
    <button
      class="bg-transparent inline-flex items-center rounded-lg text-sm text-gray-400 absolute right-2.5 top-3 ml-auto p-1.5 hover:bg-gray-200 hover:text-gray-900 dark:hover:bg-gray-800 dark:hover:text-white"
      data-testid="modal-close"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="h-5 w-5"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <div
      class="px-6 py-6 lg:px-8"
    >
      <h3
        class="text-xl font-medium text-gray-900 dark:text-white mb-3"
      >
        Sign in to our platform
      </h3>
      <form
        class="space-y-6"
        data-testid="login-form"
      >
        <div>
          <label
            class="text-sm font-medium text-gray-900 dark:text-white "
            for="email"
          >
            Your email
          </label>
          <input
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-600 dark:border-gray-500 dark:placeholder-gray-400 dark:text-white"
            data-testid=""
            name="email"
            placeholder="name@company.com"
            required=""
            type="email"
          />
        </div>
        <div>
          <label
            class="text-sm font-medium text-gray-900 dark:text-white "
            for="password"
          >
            Your password
          </label>
          <input
            class="bg-gray-50 border border-gray-300 text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 dark:bg-gray-600 dark:border-gray-500 dark:placeholder-gray-400 dark:text-white"
            data-testid=""
            name="password"
            placeholder="••••••••"
            required=""
            type="password"
          />
        </div>
        <div
          class="flex justify-between"
        >
          <div
            class="flex items-start"
          >
            <div
              class="flex h-5 items-center"
            >
              <input
                class="w-4 h-4 border border-gray-300 rounded bg-gray-50 focus:ring-3 focus:ring-blue-300 dark:bg-gray-600 dark:border-gray-500 dark:focus:ring-blue-600 dark:ring-offset-gray-800 dark:focus:ring-offset-gray-800"
                data-testid=""
                id="remember-me"
                required=""
                type="checkbox"
              />
            </div>
            <label
              class="text-sm font-medium text-gray-900 dark:text-white ml-1"
              for="remember"
            >
              Remember me
            </label>
          </div>
          <a
            class="text-sm text-blue-700 hover:underline dark:text-blue-500"
            href="src/Components/Modal/View#"
          >
            Lost Password?
          </a>
        </div>
        <button
          class="rounded-lg font-medium text-base text-center px-5 py-2.5 text-white bg-blue-700 hover:bg-blue-800 w-full"
          type="submit"
        >
          Login to your account
        </button>
        <div
          class="text-sm font-medium text-gray-500 dark:text-gray-300"
        >
          Not registered?
          <a
            class="text-blue-700 hover:underline dark:text-blue-500"
            href="src/Components/Modal/View#"
          >
            Create account
          </a>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
